voting/GeneticAlgMain.java

```diff
@@ -32,15 +38,50 @@ private void Reproduce() {
 //        System.out.println(sorted);
         ArrayList<Bundle> newPopulation = new ArrayList<>();
         // Keep best 2/3rds
-        for(int i = 0; i < k; i++) {
+        for(int i = 0; i <= k; i++) {
             newPopulation.add(sorted.get(i).getKey());
         }
-        ArrayList<Integer> parents = new ArrayList<>(k);
-        System.out.println(parents);
-        for(int i = 0; i < PopulationSize-k; i++) {
+        // Cut out last one third
+//        sorted = sorted.subList(0, k);
+        // TODO this needs to not be hard coded, match line in run vote
+        Random rnd = new Random();
+        rnd.setSeed(111);
+        ArrayList<Integer> options = new ArrayList<>();
+        for(int i = 0; i <= k; i++) {
+            options.add(i);
+        }
+        for(int i = 0; i < k/2; i++) {
+            // get one parent, remove it
+            int parentIndex = options.get(rnd.nextInt(options.size()));
+            Bundle parent1 = sorted.get(parentIndex).getKey();
+            options.remove((Integer) parentIndex);
+
+            parentIndex = options.get(rnd.nextInt(options.size()));
+            Bundle parent2 = sorted.get(parentIndex).getKey();
+            options.remove((Integer) parentIndex);
 
+            newPopulation.add(combineParents(parent1, parent2));
         }
+        this.Population = newPopulation;
     }
+    private Bundle combineParents(Bundle parent1, Bundle parent2) {
+        // Get union list
+        ArrayList<Item> unionedShows = new ArrayList<>(parent1.Bundle);
+        for (Item show : parent2.Bundle) {
+            if (!unionedShows.contains(show)) unionedShows.add(show);
+        }
+        ArrayList<Item> childBundle = new ArrayList<>();
+        while (childBundle.size() < parent1.Bundle.size()) {
+            // Get a random index and try to add it
+            int indexToTry = this.RMD.nextInt(unionedShows.size());
+            if (!childBundle.contains(unionedShows.get(indexToTry))) {
+                childBundle.add(unionedShows.get(indexToTry));
+            }
+        }
+        // TODO This needs to be a better name, but I can't remember how we decided to do that
+        return new Bundle(childBundle, "0");
```

Comment:
This line is causing the issue. Java stores values in the hashmap in the location of the hashcode and since we are overloading hashcode to be the name it's looking at that location and finding that it's not equal and thus not finding it. We need the name to be unique to each bundle for our code to work.

Reply:
never mind after running the code I think what is happening is I'm assuming that population only contains the bundles that were in the vote, but I think it also includes the reproduced bundles that were not voted on